File: python/abc298/d.py
from collections import deque

MOD = 998244353

q = int(input())

start = 0
stop = 1
state_stack = deque([1])
ans = (0, 1, 1, 1)

for _ in range(q):
    order = input()
    if order[0] == "1":
        x = int(order.split()[1])
        stop += 1
        state_stack.append(x)
    elif order[0] == "2":
        start += 1
    elif order[0] == "3":
        pre_start, pre_stop, printed_ans, pre_ans = ans
        if pre_start == start and pre_stop == stop:
            print(printed_ans)
            continue
        ans_up = pre_ans
        if pre_start != start:
            ans_up %= 10 ** (pre_stop - start)
        ans_up *= 10 ** (stop - pre_stop)
        ans_down = 0
        count = 1
        # reverse
        for i in reversed(range(pre_stop, stop)):
            ans_down += state_stack[i] * count
            count *= 10
        true_ans = ans_up + ans_down
        ans = (start, stop, true_ans, true_ans % MOD)
        print(ans[3])
    else:
        raise ValueError("invalid order")

File: rust/abc298/src/d.rs
#[allow(unused_imports)]
use proconio::{
    fastout, input
};

// I have to care overflow...
// O(q) 
fn main() {
	let module = 998244353;
	input! {
		q: usize,
	}
	let mut start = 0;
	let mut stop = 1; // not included
	let mut state_stack = vec![1];
	let mut ans = (0, 1, 1, 1); // (start, stop, ans, ans % module)
	// O(2q) = O(q)
	for _ in 0..q{
		input! { order: usize }
		match order {
			1 => {
				input! { x: usize }
				stop += 1;
				state_stack.push(x);
			}
			2 => {
				start += 1;
			}
			3 => {
				let (pre_start, pre_stop, printed_ans, pre_ans) = ans;
				if pre_start == start && pre_stop == stop{
					println!("{}", printed_ans);
					continue;
				}
				let mut ans_up = pre_ans;
				if pre_start != start{
					ans_up %= 10_i32.pow((pre_stop - start) as u32);
				}
				ans_up *= 10_i32.pow((stop - pre_stop) as u32);
				let mut ans_down = 0;
				let mut count = 1;
				// O(stop - pre_stop) means O(q) in total
				for i in (pre_stop..stop).rev(){
					ans_down += state_stack[i] * count;
					count *= 10;
				}
				let true_ans = ans_up + ans_down as i32;
				ans = (start, stop, true_ans, true_ans % module);
				println!("{}", ans.3);
			}
			_ => {
				panic!("invalid order");
			}
		}
	}
}
